### PR TITLE
Transition ClusterStatus from IntEnum to a class based state machine

### PR DESCRIPTION
**Pull request by kotfic**
@cjh1 Still need to clean up a few tests,  but this is ready to start reviewing.  It touches quite a few pieces of code that i'm only kind of familiar with so a close look would be appreciated!

### CODE REVIEW

File: cumulus/constants.py [hunk 1]
```diff
@@ -68,22 +68,103 @@ class JobQueueState:
     ERROR = 'error'
 
 
-@six.python_2_unicode_compatible
```
Comment (cjh1) [inline cumulus/constants.py]: I think we want to keep the decorator

File: cumulus/testing/__init__.py [hunk 2]
```diff
@@ -29,19 +29,8 @@ class AssertCallsMixin(object):
...
-    def _int_enum_to_string(self, data):
-        if isinstance(data, collections.Mapping):
-            for k, v in six.iteritems(data):
-                if isinstance(v, collections.Mapping):
-                    self._int_enum_to_string(v)
-                elif isinstance(v, IntEnum):
-                    data[k] = str(v)
-        elif isinstance(data, (list, tuple)):
-            for v in data:
-                self._int_enum_to_string(v)
```
Comment (cjh1) [inline cumulus/testing/__init__.py]: Excellent!

File: girder/cumulus/plugin_tests/cluster_test.py [hunk 3]
```diff
@@ -358,16 +358,45 @@ def test_update_traditional(self, generate_key):
...
+                        },
+                        u'host': u'myhost',
+                        u'ssh': {
+                            u'user': u'myuser',
+                            u'key': cluster_id,
+                            u'publicKey': self._valid_key}
+                    }, u'name': u'test'
+        }
+
         self.assertEqual(
```
Comment (cjh1) [inline girder/cumulus/plugin_tests/cluster_test.py]: Is this just a formatting change?

File: girder/cumulus/plugin_tests/cluster_test.py [hunk 4]
```diff
@@ -579,11 +610,15 @@ def test_terminate(self, terminate_cluster):
         status_body = {
             'status': 'running'
         }
+        r = self.request('/clusters/%s' %
+                         str(cluster_id),
+                         type='application/json', method='PATCH',
+                         body=json.dumps(status_body), user=self._user)
```
Comment (cjh1) [inline girder/cumulus/plugin_tests/cluster_test.py]: Why was this added?

File: girder/cumulus/server/cluster.py [hunk 5]
```diff
@@ -332,14 +334,11 @@ def update(self, id, params):
...
         if 'status' in body:
-            try:
-                # For now only do for ansible clusters
-                if cluster['type'] in [ClusterType.ANSIBLE, ClusterType.EC2]:
-                    cluster['status'] = ClusterStatus[body['status']]
-                else:
-                    cluster['status'] = body['status']
-            except (ValueError, KeyError):
+            if ClusterStatus.valid(body['status']):
                 cluster['status'] = body['status']
```
Comment (cjh1) [inline girder/cumulus/server/cluster.py]: Nice to have validation of the status value here.

File: girder/cumulus/server/cluster.py [hunk 6]
```diff
@@ -470,8 +460,7 @@ def submit_job(self, id, jobId, params):
         if not cluster:
             raise RestException('Cluster not found.', code=404)
 
-        if cluster['status'] != 'running' and \
-           cluster['status'] != ClusterStatus.running:
```
Comment (cjh1) [inline girder/cumulus/server/cluster.py]: Ug, that was horrible!

File: girder/cumulus/server/utility/cluster_adapters.py [hunk 7]
```diff
@@ -192,18 +202,17 @@ def provision(self):
         cumulus.ansible.tasks.cluster.provision_cluster \
             .delay(playbook, self.cluster, profile, secret_key,
                    playbook_params,
-                   girder_token, log_write_url, 'provisioned')
+                   girder_token, log_write_url, 'running')
```
Comment (cjh1) [inline girder/cumulus/server/utility/cluster_adapters.py]: Should this be ClusterStatus.RUNNING?

File: girder/cumulus/plugin_tests/cluster_test.py (hunk 3 above)
Comment (kotfic) [inline girder/cumulus/plugin_tests/cluster_test.py]: Yes just a format change

File: girder/cumulus/plugin_tests/cluster_test.py (hunk 4 above)
Comment (kotfic) [inline girder/cumulus/plugin_tests/cluster_test.py]: We can't terminate a 'created' cluster so we patch the status of the cluster to 'running' so the validation works.
